packages/config/src/files.ts

```diff
@@ -0,0 +1,189 @@
+import { existsSync } from 'fs'
+import { resolve, relative, parse, join } from 'path'
+
+import { getProperty, setProperty, deleteProperty } from 'dot-prop'
+
+import { throwUserError } from './error.js'
+import { mergeConfigs } from './merge.js'
+import { isTruthy } from './utils/remove_falsy.js'
+
+/**
+ * We allow paths in configuration file to start with /
+ * In that case, those are actually relative paths not absolute.
+ */
+const LEADING_SLASH_REGEXP = /^\/+/
+
+/**
+ * All file paths in the configuration file are relative to `buildDir`
+ * (if `baseRelDir` is `true`).
+ */
+const FILE_PATH_CONFIG_PROPS = [
+  'functionsDirectory',
+  'functions.*.deno_import_map',
+  'build.publish',
+  'build.edge_functions',
+]
+
+/**
+ * Make configuration paths relative to `buildDir` and converts them to
+ * absolute paths
+ */
+export const resolveConfigPaths = function (options: {
+  config: $TSFixMe
+  repositoryRoot: string
+  buildDir: string
+  baseRelDir?: boolean
+  packagePath?: string
+}) {
+  const baseRel = options.baseRelDir ? options.buildDir : options.repositoryRoot
+  const config = resolvePaths({
+    config: options.config,
+    packagePath: options.packagePath,
+    propNames: FILE_PATH_CONFIG_PROPS,
+    baseRel,
+    repositoryRoot: options.repositoryRoot,
+  })
+  return addDefaultPaths({
+    config,
+    repositoryRoot: options.repositoryRoot,
+    baseRel,
+    packagePath: options.packagePath,
+  })
+}
+
+const resolvePaths = function ({
+  config,
+  propNames,
+  baseRel,
+  packagePath,
+  repositoryRoot,
+}: {
+  config: $TSFixMe
+  propNames: string[]
+  baseRel: string
+  repositoryRoot: string
+  packagePath?: string
+}) {
+  return propNames.reduce(
+    (configA, propName) => resolvePathProp(configA, propName, baseRel, repositoryRoot, packagePath),
+    config,
+  )
+}
+
+const resolvePathProp = function (
+  config: $TSFixMe,
+  propName: string,
+  baseRel: string,
+  repositoryRoot: string,
+  packagePath?: string,
+) {
+  const path = getProperty(config, propName) as string | undefined
+
+  if (!isTruthy(path)) {
+    deleteProperty(config, propName)
+    return config
+  }
+
+  return setProperty(config, propName, resolvePath(repositoryRoot, baseRel, path, propName, packagePath))
+}
+
+export const resolvePath = (
+  repositoryRoot: string,
+  baseRel: string,
+  originalPath: string,
+  propName: string,
+  // @ts-expect-error depends on the survey outcome see comment below
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  packagePath?: string,
+) => {
+  if (!isTruthy(originalPath)) {
+    return
+  }
+
+  const path = originalPath.replace(LEADING_SLASH_REGEXP, '')
+  // TODO: Based on survey outcome: https://netlify.slack.com/archives/C05556LEX28/p1691423437855069
+  // If we like option B then:
+  // const pathA = resolve(baseRel, packagePath || '', path)
```

Comment:
This would be `Option B` in the discussion: https://netlify.slack.com/archives/C05556LEX28/p1691423437855069